### DISCUSSION TITLE
Question about LM adapted T5 dataset

### DISCUSSION THREAD

**Discussion by YU-Anthony** [NVIDIA-NeMo/NeMo#3940] [Q&A]
Hi all,

Thank you for your effort for these work. I have a little question about the split point of the input source. As mentioned in T5 paper, they choose a random point to split it into prefix and target portions. However, you set a specific length to split text sequence into two parts as shown in the following code

https://github.com/NVIDIA/NeMo/blob/49183b7c59c3b4243671d5eb4d3f202a4e4739ce/nemo/collections/nlp/data/language_modeling/megatron/lm_adapted_t5_dataset.py#L49-L52

I am curious about this part. Could you please explain why this set is different from T5 paper? Thank you in advance.

**Comment by MaximumEntropy** [comment]
Hi, great question! 

(a) I picked a fixed spot to split to minimize the amount of padding within a batch. If each example within a batch has a different split point, there might be a fair bit of padding within a batch. 

(b) My goal was to implement the LM Adapted T5 model described in https://arxiv.org/abs/2104.08691 rather than the original paper. The paper however lacks a lot of detail on how this fine-tuning was done so I took my best guess.

(c) We can definitely support random split points at the expense of padding. Happy to review a PR if you have time to work on this, if not I'll try and get to it in the coming weeks :)

**Comment by YU-Anthony** [comment]
Thank you for your reply! So, you set the sequence length of prefix to 128 right? Have you ever tried different settings? 
https://github.com/NVIDIA/NeMo/blob/bdb561cf744b3fc74e67c1932232dd1ce0740c42/Jenkinsfile#L2254

> Hi, great question!
> 
> (a) I picked a fixed spot to split to minimize the amount of padding within a batch. If each example within a batch has a different split point, there might be a fair bit of padding within a batch.
> 
> (b) My goal was to implement the LM Adapted T5 model described in https://arxiv.org/abs/2104.08691 rather than the original paper. The paper however lacks a lot of detail on how this fine-tuning was done so I took my best guess.
> 
> (c) We can definitely support random split points at the expense of padding. Happy to review a PR if you have time to work on this, if not I'll try and get to it in the coming weeks :)

**Comment by MaximumEntropy** [comment]
That is just in our CI tests. Our defaults while training models are 512 encoder sequence length and 128 for the decoder. Internally, we've trained models with 512/128 and 1024/256. Take a look at your YAML config to things that you can configure while pre-training (https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/conf/megatron_t5_config.yaml).

**Comment by YU-Anthony** [comment]
That helps me a lot ! I'll do some experiments on these settings.
